implement a function to materialize a copy-on-write storage

dagitses opened the pull request:
implement a function to materialize a copy-on-write storage

Summary:
This is triggered if the storage is copy-on-write whenever mutable
access to a DataPtr is given.

Test Plan: 100% code coverage

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/pytorch/pull/100820).
* #100821
* __->__ #100820

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

ezyang commented on c10/core/impl/cow/materialize.cpp:
```diff
@@ -0,0 +1,54 @@
+    // We don't need to consume the result, it's just a shared lock
+    // ensuring that the data will remain while we copy it.
+    new_data_ptr = storage.allocator()->allocate(storage.nbytes());
+    std::memcpy(new_data_ptr->get(), data_ptr.get(), storage.nbytes());
```
This is wrong. The data pointer may not point to CPU memory, it may be CUDA memory. We may need to teach allocators to know how to do memcpy's.

dagitses replied:
OK, I added two methods to allocator: non-virtual `clone` which delegates to `virtual copy_data`. `copy_data` defaults to `std::memcpy` and I override it for the cuda allocator. Is that a sensible path?

dagitses commented on c10/core/impl/cow/materialize.cpp:
```diff
@@ -0,0 +1,54 @@
+
+auto C10_API cow::materialize(StorageImpl& storage) -> void {
+  at::DataPtr& data_ptr = storage.mutable_data_ptr();
+  TORCH_INTERNAL_ASSERT(data_ptr.get_deleter() == cow::delete_context);
```
Fair. I don't like to rely on undocumented functions, but I documented the cast_context function in the previous PR now, so it's ok to do.

dagitses committed:
implement a function to materialize a copy-on-write storage

Summary:
This is triggered if the storage is copy-on-write whenever mutable
access to a DataPtr is given.

Test Plan: 100% code coverage

kurtamohler commented:
I'm continuing to work on COW storages, and the next step is to get this PR merged or submit a new PR to replace it. One issue is, as Edward pointed out already, we need the copy to work for any device. But it would be nice if we could just reuse the device-agnostic copy that we already have, [`at::native::copy_impl`](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Copy.cpp#L134). However, this PR puts the materialize function in `c10`, so we don't have access to `at` functions.

There are at least three potential options to consider:

1. Reimplement device-agnostic copy in `c10`, like dagitses decided to do. If this turns out to be a good solution, then great. But it seems worth considering alternatives that would avoid adding a second implementation of a device-agnostic copy. This PR doesn't support all the devices that `at::native::copy_impl` supports yet, so if we go this route, the missing ones will need to be added.

2. Move the underlying implementation of device-agnostic copy to `c10`. `at::native::copy_impl` would call into the `c10` implementation. However, there is at least one problem with this idea that probably makes it infeasible or impractical. `copy_impl` uses TensorIterator to enable broadcasting two tensors that aren't the same shape. The `c10` device-agnostic copy would just be a contiguous data copy. I don't think it would really make sense to use a `c10` contiguous data copy to implement the broadcasting copy.

3. Move the materialize function to `at` so that it can use `at::native::copy_impl`. Obviously we would no longer be able to call the materialize function from `c10::StorageImpl::mutable_data` or any other `c10` context in this case. We'd also probably have to throw an error if `StorageImpl::mutable_data` is called on a COW storage. Any `at` operation that writes to a COW storage would have to call the materialize function at some point before the write happens. I don't know exactly how many call sites this would actually amount to, but it might not be a lot. I'm not 100% sure if this solution will be feasible or not--if there are any calls to `StorageImpl::mutable_data` from within `c10` for which we cannot move the materialize function call up to `at` or another context that has access to `at`, then it won't be feasible

After thinking through these options a little bit, I think option 1 is the best after all. Option 3 involves more unknowns and might end up being more complicated even though it avoids implementing a second device-agnostic copy. @ezyang , I'm curious if you agree or have any other suggestions

ezyang commented:
(1) is fine by me. We have some precedent for device specific dispatch on DeviceGuardImpl and adding a copy doesn't seem like the end of the world to me.

kurtamohler commented:
Replaced by #113396
